File: src/main/java/org/voegtle/weatherwidget/preferences/WeatherPreferenceFragment.java
package org.voegtle.weatherwidget.preferences;

import android.content.pm.PackageManager;
import android.os.Bundle;
import android.preference.PreferenceCategory;
import android.preference.PreferenceFragment;
import android.preference.PreferenceScreen;
import org.voegtle.weatherwidget.R;
import org.voegtle.weatherwidget.util.ContextUtil;

public class WeatherPreferenceFragment extends PreferenceFragment {
  private static final String PLACEHOLDER_VERSION = "{v}";
  private static final String PLACEHOLDER_BUILD = "{b}";
  private int buildNumber;
  private String versionName;

  @Override
  public void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    addPreferencesFromResource(R.xml.preferences);

    readVersionInformationFromAndroidManifest();
    addBuildInformation(getVersionNotice());

  }

  private void addBuildInformation(PreferenceScreen appInfoScreen) {
    String versionInfo = appInfoScreen.getSummary().toString();
    versionInfo = versionInfo.replace(PLACEHOLDER_VERSION, versionName);
    versionInfo = versionInfo.replace(PLACEHOLDER_BUILD, Integer.toString(buildNumber));
    appInfoScreen.setSummary(versionInfo);

    appInfoScreen.getPreference(0).setTitle(versionInfo);
  }


  private PreferenceScreen getVersionNotice() {
    int lastIndex = getPreferenceScreen().getPreferenceCount() - 1;

    PreferenceCategory versionCategory = (PreferenceCategory) getPreferenceScreen().getPreference(lastIndex);
    return (PreferenceScreen) versionCategory.getPreference(0);
  }

  private void readVersionInformationFromAndroidManifest() {
    try {
      buildNumber = ContextUtil.getBuildNumber(getActivity());
      versionName = ContextUtil.getVersion(getActivity());
    } catch (Exception ignored) {
    }
  }

}

File: app/src/main/java/org/voegtle/weatherwidget/preferences/WeatherPreferenceFragment.kt
package org.voegtle.weatherwidget.preferences

import android.os.Bundle
import androidx.fragment.app.Fragment
import androidx.preference.Preference
import androidx.preference.PreferenceFragmentCompat
import androidx.preference.PreferenceScreen
import org.voegtle.weatherwidget.R
import org.voegtle.weatherwidget.util.ContextUtil

class WeatherPreferenceFragment : PreferenceFragmentCompat(), PreferenceFragmentCompat.OnPreferenceStartScreenCallback {
    private val PLACEHOLDER_VERSION = "{v}"
    private val PLACEHOLDER_BUILD = "{b}"

    private val VERSION_INFO_KEY = "version"

    override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {
        setPreferencesFromResource(R.xml.preferences, rootKey)

        var versionName = ""
        var buildNumber = 0

        try {
            val activity = requireActivity()
            versionName = ContextUtil.getVersion(activity)
            buildNumber = ContextUtil.getBuildNumber(activity)
        } catch (e: IllegalStateException) {
        }

        val versionPreferenceItem: Preference? = findPreference(VERSION_INFO_KEY)

        versionPreferenceItem?.let { pref ->
            val currentSummary = pref.summary?.toString() ?: "Version {v} (Build {b})"
            val versionInfoText = currentSummary
                .replace(PLACEHOLDER_VERSION, versionName)
                .replace(PLACEHOLDER_BUILD, buildNumber.toString())

            pref.summary = versionInfoText

            if (pref is PreferenceScreen && pref.key == VERSION_INFO_KEY) {
                if (pref.preferenceCount > 0) {
                    val firstChild: Preference? = pref.getPreference(0)
                    firstChild?.title = versionInfoText
                }
            }
        }
    }

    override fun onPreferenceStartScreen(
        caller: PreferenceFragmentCompat,
        screen: PreferenceScreen
    ): Boolean {
        val fragment: Fragment = WeatherPreferenceFragment()
        val args = Bundle()
        args.putString(ARG_PREFERENCE_ROOT, screen.key)
        fragment.arguments = args

        parentFragmentManager.beginTransaction()
            .replace(R.id.fragment_container, fragment)
            .addToBackStack(screen.key)
            .commit()
        return true
    }
}
